articles/tuples-data-structure-python/index.md

```diff
@@ -0,0 +1,307 @@
+
+```py
+tuple1 = (0, 1, 2, 3)
+
+print(tuple1[0]) # Output: 0
+
+print(tuple1[1]) # Output: 1
+```
+
+This tuple follows zero indexing.
+
+![Tuple Positive Indexing](/engineering-education/tuples-data-structure-python/tuple-positive-index.png)
+
+_Tuple Positive Indexing: Source -- [GeeksforGeeks](https://www.geeksforgeeks.org/python-tuples/)_
+
+##### Negative Indexing
+
+Similar to lists, we can also use negative indexing on a tuple. Therefore, '-1' refers to the Nth element of a tuple, -2 refers to the (N-1)th element, and so on (where N is the length of the tuple).
+
+```py
+tuple1 = (30, 40, 50, 60)
+
+print(tuple1[-1]) # Output: 60
+
+print(tuple1[-3]) # Output: 40
+```
+
+![Tuple Negative Indexing](/engineering-education/tuples-data-structure-python/tuple-negative-index.png)
+
+_Tuple Negative Indexing_
+
+##### Slicing
+
+In Python, slicing is used to return a range of values. Like lists, tuples can also be sliced.
+
+```py
+tuple1 = (1, 2, 3, 5, 8, 13)
+
+print(tuple1[0:3]) # Output: (1, 2, 3)
+
+print(tuple1[4:]) # Output: (8, 13)
+```
+
+As per the examples shown above, if we slice a range of [a : b), it would return from tuple index a to tuple index (b - 1). For more tricks on python slicing, check out [this page](https://stackoverflow.com/questions/509211/understanding-slice-notation).
+
+#### Modifying Tuples
+
+Tuples are immutable. For example:
+
+```py
+tuple1 = (2000, 3000, 4000)
+
+tuple1[1] = 1000
+```
+
+If we execute the above code, the python interpreter throws the following error:
+
+```bash
+Traceback (most recent call last):
+    File "main.py", line 3 in <module>
+        tuple1[1] = 1000
+TypeError: 'tuple' object does not support item assignment
+```
+
+This is because a tuple is designed to be immutable. However, we can change a tuple that contains mutable objects. For example, let us take a tuple of lists.
+
+```py
+tuple1 = ([10, 20], [30, 40], [50, 60])
+
+tuple1[1][0] = 70
+
+print(tuple1) # Output: ([10, 20], [70, 40], [50, 60])
+```
+
+This works perfectly because we are modifying the list within a tuple (which is mutable). We can also create new tuples from existing ones.
+
+```py
+tuple1 = ([10, 20], [30, 40], [50, 60])
+tuple2 = ([100, 200], [300, 400])
+
+# Creating a new tuple from tuple1 and tuple2
+tuple3 = tuple1 + tuple2
+
+print(tuple3) # Output: ([10, 20], [30, 40], [50, 60], [100, 200], [300, 400])
+```
+
+#### Tuple Methods
+
+Tuples have the following in-built methods that make them extremely powerful:
+
+- **`cmp(tuple1, tuple2)`**
+- **`len(tuple)`**
+- **`min(tuple)`**
+- **`max(tuple)`**
+- **`tuple(list)`**
+- **`t.count(el)`**
+- **`t.index(el)`**
+
+##### cmp(tuple1, tuple2)
+
+_Note: The cmp() method existed in python2. It wasn't included in python3. Therefore we define our own compare method._
+
+The compare method analyses two tuples element by element. It compares them and returns the following:
```

Comment:
This is an interesting example. Can you consider the example of (100,300) and (200,100)? What would the output be?

Reply:
The output would be 1.
This is because the tuple comparison is done left to right. When tuple2[0] > tuple1[0], no further comparisons are made and the output is returned as zero. This is how the cmp() method works in Python.
I've added the given example in the recent commit 💯